### ISSUE TITLE
Questions about big dataset processing or split reading

### ISSUE THREAD

**Issue by odymit** [THUDM/cogdl#366]
## ❓ Questions & Help

Hi, I am a beginer in gnn, thanks for the convenient tool COGDL open sourced by THUDM.

## Question Description
I implemented a customized Dataset for graph classification task. I have proprecessed the data, and when I am loading the dataset, failed with memory exceeded.

## Detail Info
Here is my code example, and memory usage:
```Python
import os
import torch
from cogdl.data import Graph
from cogdl.datasets import GraphDataset

class MyGraphDataset(GraphDataset):
    def __init__(self, path="data.pt"):
        self.path = path
        super(MyGraphDataset, self).__init__(path, metric="accuracy")

    def process(self):
        # Load and preprocess data
        graphs = []
        # about 5000 data in the data_dir
        for file in os.listdir(data_dir):
            data = torch.load(file)
            edges = data['edges']
            # node feature is pretty big, a 250 000 dim tensor in float
            x = data['x']
            label = data['y']
            graphs.append(Graph(edge_index=edges, x=x, y=label))
        return graphs
if __name__ == "__main__":
    dataset = MyGraphDataset()
    experiment(model="gin", dataset=dataset)
```
![86275b250320fdbb2a8d6798b69ffa3](https://user-images.githubusercontent.com/15167656/181782488-82ee699b-3265-4720-a4dc-c68c0d71e59f.png)

The usage pic above is capture when all data loaded about 30%.
## Concrete Question
I am wandering if there are ways to get a `MyGraphDataset` instance or other ways to train using `experiment` function, and I also tried several ways to get it done. Such as:
-  `MyGraphDataset` instance could be split reading or a batch mode, such as yield
- trying to figure out how the data.pt produce process and compress way
- read limited data by monitoring the memory usage, break the reading loop and return graphs
- refer to other implementation, I am reading the source code currently

So, if there ways to train a model on graph classification task using big single graph when dataset size is pretty big?

**Comment by hhtttyy** [comment]
i think you should concat all  tensors as input to create **Graph** in process function ,and tutorial document said this function should return a Graph data structure.i have a bigger data than you and already processed and stored in disk. But,i have problem in traing my data in unsup_graphsage model by experiment api ,i'm still working on it

**Comment by odymit** [comment]
> i think you should concat all tensors as input to create **Graph** in process function ,and tutorial document said this function should return a Graph data structure.
Thanks for attention, sorry about the example code, i will correct it laterly, in my exp code:
- [512,512] is the original shape, and i have reshaped it for Graph creation
- graphs is the parameter returned by the process function

>i have a bigger data than you and already processed and stored in disk. 

you mean the process of `Dataset` class is done, and you have got the `data.pt` in disk?

>But,i have problem in traing my data in unsup_graphsage model by experiment api ,i'm still working on it

you have same problem about getting a `Dataset` instance or the training on big dataset only?

**Comment by hhtttyy** [comment]
1.YES i got data.pt from my data ,but in my code i concat all my node feature `shape (dim,)` to create only one graph rather than every node feature to create own graph such in your code , maybe different task.
2. In traing ,and i propose #365  
i'm also a begginer , this answer may not help you 🙈

**Comment by cenyk1230** [comment]
Hi @odymit,

Thanks for your interest in our toolkit. 
From your description, you have 5000 graphs for graph classification. Each graph has 512(?) nodes with 512-dim(?) features. Therefore, it will have 5000*512*512=1.3B floats. Moreover, each graph has edges between 512 nodes. What is the average number of edges of these graphs? Let me know if I misunderstand something.

**Comment by odymit** [comment]
Thanks for you rely @cenyk1230 ,
But that's not the key point, I would like to known if there is a solution using a dataset which size is bigger than server memory for training? 

When processing the `MyGraphDataset`, now the `graphs` value only returned after the whole dataset loaded. Whatever the size each graph is, when the whole dataset is bigger than the memory, the dataset can't  be loaded as a `MyGraphDataset`. And I can't get a `MyGraphDataset` instance for the following experiment function or other things.

By the way, when I am using the experiment function in limited dataset size(got the `MyGraphDataset` instance, and `data.pt`), it seems like using the same way loading dataset all in the memory, and then the training procedure starts.

**Comment by cenyk1230** [comment]
Hi @odymit,

Currently, we assume that all data need to be loaded into the memory at the beginning of the training. Therefore, CogDL only supports the dataset where the size is smaller than the memory. I'm sorry for that. But we will discuss and explore how to deal with this case in the future.

**Comment by odymit** [comment]
@cenyk1230 , thanks very much.

Also, I am glad to contribute to CogDL if needed.

**Comment by odymit** [comment]
Use dataloader.
